src/util/pingproperties.h

```diff
@@ -0,0 +1,27 @@
+#pragma once
+
+#include <QtGlobal>
+
+/** Calls the correct operator == in the P_PROPERTY macro */
+template<typename T> bool pingIsSame(const T& a, const T& b)
```

Comment:
how is this different from the cpp ones ?
Like, why this definition is here and the others in the cpp ?

Reply:
because this is a generic template. What happens is when you have a non specialized template, the code can't be in the .cpp as the code will be generated during compilation. 

You can have three possibilities with templates:
- No specializations at all: all of the code must be in the header
- Partial specialization: if at least *one* of the template parameters is a template generic type, all the code must be in the header.
- Full specialization: if the code is fully specialized (no template parameter at all), the code cannot be in the header as multiple definition errors would occour in the link time.


The first one defines a template, the other two fully specializes it.